File: AppKartingRM/src/main/java/com/kartingRM/AppKartingRM/services/ReservaService.java
package com.kartingRM.AppKartingRM.services;

import com.kartingRM.AppKartingRM.entities.*;
import com.kartingRM.AppKartingRM.repositories.ReservaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.*;

@Service
public class ReservaService {

    @Autowired
    private ReservaRepository reservaRepository;
    @Autowired
    private PlanService planService;
    @Autowired
    private ClienteService clienteService;


    // Obtener todas las reservas
    public List<ReservaEntity> getReservas(){
        return reservaRepository.findAll();
    }

    // Obtener reserva segun id
    public ReservaEntity getReservaById(Long id){
        return reservaRepository.findById(id).get();
    }

    // Obtener todas las reservas segun id de un cliente reservante
    public List<ReservaEntity> getReservasByClienteId(Long clienteId) {
        return reservaRepository.findByReservanteId(clienteId);
    }

    // Obtener reservas existentes entre dos horas de un dia
    public List<ReservaEntity> getReservasSolapadasByHoraInicio(LocalDate fecha, LocalTime horaInicio, LocalTime horaFinal){
        return reservaRepository.findReservasExistentesEnTiempo(fecha, horaInicio, horaFinal);
    }

    // Obtener integrantes de reserva segun id
    public List<ClienteEntity> getIntegrantesById(Long id) {
        ReservaEntity reserva = reservaRepository.findById(id).get();   // Obtengo reserva
        // Entrego los integrantes
        return reserva.getIntegrantes();
    }

    // Funcion que ingresa una reserva a la base de datos
    // Considera que no existan reservas en el horario nuevo, ademas de que se ingresa automaticamente
    //    la hora de fin segun el tiempo del plan
    public ReservaEntity createReserva(ReservaEntity reserva, Long idCliente, Long idPlan, Boolean esFeriado){
        PlanEntity plan = planService.getPlanById(idPlan);
        ClienteEntity cliente = clienteService.getClienteById(idCliente);
        // Caso parametros validos de cliente, plan y hora de inicio
        if (cliente != null && plan != null && reserva.getHoraInicio() != null) {
            reserva.setReservante(cliente);
            reserva.setPlan(plan);

            // Determinar la hora final sumando los minutos del plan a la hora inicial
            LocalTime horaInicio = reserva.getHoraInicio();
            LocalTime horaFinalCalculada = reserva.getHoraInicio().plusMinutes(reserva.getPlan().getDuracionTotal());
            reserva.setHoraFin(horaFinalCalculada);

            // Verificar que no existan reservas existentes en el horario determinado
            List<ReservaEntity> reservasExistentes = getReservasSolapadasByHoraInicio(
                    reserva.getFecha(), reserva.getHoraInicio(), reserva.getHoraFin());

            // Si existen reservas en el horario, no sea crea la nueva reserva
            if (!reservasExistentes.isEmpty()){
                throw new IllegalStateException("Ya existe una reserva con ese horario");
            }

            // Verificar Horario de inicio y fin validos.
            //  Lunes a Viernes: 14:00 a 22:00
            //  Sabados, Domingos, Feriados: 10:00 a 22:00
            boolean esFinDeSemana = reserva.getFecha().getDayOfWeek().getValue() >= 6;// Analisis si es fin de semana o no
            if (esFeriado || esFinDeSemana){// Horario fin de semana o feriado
                // Horario antes o despues del horario de servicio
                if (horaInicio.isBefore(LocalTime.of(10,00,00)) || horaFinalCalculada.isAfter(LocalTime.of(22,00,00))) {
                    throw new IllegalStateException("Horario incorrecto. Domingos, sábados y feriados: 10:00 a 22:00");
                }
            } else { // Horario semana
                if (horaInicio.isBefore(LocalTime.of(14,00,00)) || horaFinalCalculada.isAfter(LocalTime.of(22,00,00))) {
                    throw new IllegalStateException("Horario incorrecto. Lunes a Viernes: 14:00 a 22:00");
                }
            }

            return reservaRepository.save(reserva);
        } else {
            throw new RuntimeException("Cliente no encontrado, plan no existe, o hora erronea");
        }
    }

    // Actualizar reserva
    public ReservaEntity updateReserva(Long id, ReservaEntity reserva){
        // Obtiene el objeto Reserva sin modificar
        ReservaEntity reservaOriginal = reservaRepository.findById(id).get();
        // Seteamos a los valores referenciados FK de la reserva original
        reserva.setIdReserva(id);
        reserva.setPlan(reservaOriginal.getPlan());
        reserva.setReservante(reservaOriginal.getReservante());

        // Actualizacion nuevo hora de fin, caso cambio de plan o de hora de inicio
        LocalTime nuevaHoraFin = reserva.getHoraInicio().plusMinutes(reserva.getPlan().getDuracionTotal());
        reserva.setHoraFin(nuevaHoraFin);

        // Falta verificar si cambio el tipo de plan para actualizar los tiempos o el horario para actualizar el horario final
        return reservaRepository.save(reserva);
    }

    // Actualizar cliente de una reserva
    public ReservaEntity updateClienteDeReserva(Long idReserva, Long idCliente){
        // Obtener objeto Reserva
        ReservaEntity reservaOriginal = reservaRepository.findById(idReserva).get();
        // Obtener objeto Cliente
        ClienteEntity cliente = clienteService.getClienteById(idCliente);
        reservaOriginal.setReservante(cliente);

        // Actualizacion nuevo hora de fin, caso cambio de plan o de hora de inicio
        LocalTime nuevaHoraFin = reservaOriginal.getHoraInicio().plusMinutes(reservaOriginal.getPlan().getDuracionTotal());
        reservaOriginal.setHoraFin(nuevaHoraFin);
        return reservaRepository.save(reservaOriginal);
    }

    // Actualizar plan de una reserva
    public ReservaEntity updatePlanDeReserva(Long idReserva, Long idPlan){
        // Obtener objeto Reserva
        ReservaEntity reservaOriginal = reservaRepository.findById(idReserva).get();
        // Obtener objeto Plan
        PlanEntity plan = planService.getPlanById(idPlan);
        reservaOriginal.setPlan(plan);

        // Actualizacion nuevo hora de fin, caso cambio de plan o de hora de inicio
        LocalTime nuevaHoraFin = reservaOriginal.getHoraInicio().plusMinutes(reservaOriginal.getPlan().getDuracionTotal());
        reservaOriginal.setHoraFin(nuevaHoraFin);
        return reservaRepository.save(reservaOriginal);
    }

    // Eliminar una reserva segun su id
    public boolean deleteReserva(Long id) throws Exception{
        try{
            reservaRepository.deleteById(id);
            return true;
        } catch (Exception e){
            throw new Exception(e.getMessage());
        }
    }

    // Obtiene los ingresos segun plan en un rango de tiempo
    public List<Object[]> findIngresosByVueltasAndFlexibleRange(int mesInicio, int yearInicio, int mesFin, int yearFin ){
        return reservaRepository.findIngresosByVueltasAndFlexibleRange(mesInicio, yearInicio, mesFin, yearFin);
    }

    // Obtiene los ingresos segun rango de personas en la reserva
    public List<Object[]> findIngresosByRangoPersonas(int mesInicio, int yearInicio, int mesFin, int yearFin ){
        return reservaRepository.findIngresosByRangoPersonas(mesInicio, yearInicio, mesFin, yearFin);
    }



    // Agregar un cliente a la lista de integrantes de una reserva
    public ReservaEntity agregarIntegrante(Long reservaId, Long clienteId) {
        ReservaEntity reserva = reservaRepository.findById(reservaId)
                .orElseThrow(() -> new RuntimeException("Reserva no encontrada"));

        // Caso en que se intentan agregar mas integrantes de la cantidad total de integrantes de la reserva
        if (reserva.getIntegrantes().size() >= reserva.getTotalPersonas()) {
            throw new IllegalStateException("Ya no se puede asignar mas integrantes");
        } else {
            // Caso en que todavia no se agregan todos los integrantes a la reserva
            // Obtener cliente segun id
            ClienteEntity cliente = clienteService.getClienteById(clienteId);

            // Si es que la reserva no tiene al cliente como integrante, se agrega, en caso contrario no
            if (!reserva.getIntegrantes().contains(cliente)) {
                reserva.getIntegrantes().add(cliente);
            }

            return reservaRepository.save(reserva);
        }
    }

    // Quitar un cliente segun id de la lista de integrantes de una reserva
    public ReservaEntity quitarIntegrante(Long reservaId, Long clienteId) {
        ReservaEntity reserva = reservaRepository.findById(reservaId)
                .orElseThrow(() -> new RuntimeException("Reserva no encontrada"));
        // Obtener cliente segun id
        ClienteEntity cliente = clienteService.getClienteById(clienteId);

        reserva.getIntegrantes().remove(cliente);

        return reservaRepository.save(reserva);
    }
}

File: FrontEnd KartingRM/KartingRM/src/services/reservaService.js
import axios from 'axios';

const API_URL = '/api/reservas';

// Peticion GET de todas las reservas
export const getReservas = async () => {
  try {
    const response = await axios.get(`${API_URL}/`);
    return response.data;
  } catch (error) {
    console.error('Error al obtener reservas:', error);
    throw error;
  }
};

// Peticion Get de reserva segun id
export const getReservaById = async (idReserva) => {
  try {
    const response = await axios.get(`${API_URL}/${idReserva}`);
    return response.data;
  } catch (error) {
    console.error(`Error al obtener reserva con ID ${idReserva}:`, error);
    throw error;
  }
};

// Peticion Get de reservas de un cliente segun id
export const getReservasByCliente = async (idCliente) => {
  try {
    const response = await axios.get(`${API_URL}/por-cliente/${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al obtener reservas del cliente ${idCliente}:`, error);
    throw error;
  }
};

// Peticion POST para crear reserva indicando id de cliente, id de plan, booleano que indica si 
//  es dia feriado, y el cuerpo de la reserva
export const createReserva = async (idCliente, idPlan, feriado, reservaData) => {
  try {
    const response = await axios.post(
      `${API_URL}?id_cliente=${idCliente}&id_plan=${idPlan}&feriado=${feriado}`,
      reservaData
    );
    return response.data;
  } catch (error) {
    console.error('Error al crear reservaxx:', error);
    throw error;
  }
};

// Peticion PUT para update de reserva segun cuerpo
export const updateReserva = async (idReserva, reservaData) => {
  try {
    const response = await axios.put(`${API_URL}/${idReserva}`, reservaData);
    return response.data;
  } catch (error) {
    console.error(`Error al actualizar reserva con ID ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PUT para update del cliente reservante de una reserva
export const updateClienteReserva = async (idReserva, idCliente) => {
  try {
    const response = await axios.put(`${API_URL}/${idReserva}/cliente?id_cliente=${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al actualizar cliente en reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PUT para update del plan de una reserva
export const updatePlanReserva = async (idReserva, idPlan) => {
  try {
    const response = await axios.put(`${API_URL}/${idReserva}/plan?id_plan=${idPlan}`);
    return response.data;
  } catch (error) {
    console.error(`Error al actualizar plan en reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PATCH para relacionar cliente integrante con reserva
export const addIntegrante = async (idReserva, idCliente) => {
  try {
    const response = await axios.patch(`${API_URL}/${idReserva}/integrantes/agregar/${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al agregar integrante a reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PATCH para quitar un cliente como integrante de una reserva
export const removeIntegrante = async (idReserva, idCliente) => {
  try {
    const response = await axios.patch(`${API_URL}/${idReserva}/integrantes/quitar/${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al quitar integrante de reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion DELETE para eliminar una reserva
export const deleteReserva = async (idReserva) => {
  try {
    await axios.delete(`${API_URL}/${idReserva}`);
    return idReserva;
  } catch (error) {
    console.error(`Error al eliminar reserva con ID ${idReserva}:`, error);
    throw error;
  }
};